site/src/pages/docs/client-decorator.mdx

```diff
@@ -169,6 +169,77 @@ preview because <type://ContentPreviewingClient> is evaluated first for the HTTP
 For <type://RetryingClient>, please check out 
 [`RetryingClient` with logging](/docs/client-retry#retryingclient-with-logging).
 
+## Modifying responses
+
+We can modify the responses in a decorator via wrapping the <type://HttpResponse> with <type://FilteredHttpResponse>:
+
+```java
+import com.linecorp.armeria.client.ClientBuilder;
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.common.FilteredHttpResponse;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.ResponseHeaders;
+import java.nio.charset.StandardCharsets;
+
+ClientBuilder cb = Clients.builder(...);
+cb.decorator((delegate, ctx, req) -> {
+    HttpResponse response = delegate.execute(ctx, req);
+    return new FilteredHttpResponse(response) {
+        boolean initialHttpData = true;
+        protected HttpObject filter(HttpObject obj) {
+            if (obj instanceof ResponseHeaders) {
+                // Can receive multiple ResponseHeaders whose status is informational.
+                ResponseHeaders responseHeaders = (ResponseHeaders) obj;
+                if (responseHeaders.status().isInformational()) {
+                    return obj;
+                }
+                // Add a "x-foo" header into the response.
+                return responseHeaders.toBuilder()
+                                      .add("x-foo", "foobar")
+                                      .build();
+            } else if (obj instanceof HttpData) {
+                HttpData httpData = (HttpData) obj;
+                // Example of how to insert attribute into json payload.
+                // initial '{' in response is assumed to be first byte
+                if (initialHttpData) {
+                    initialHttpData = false;
+                    byte[] ascii = "{\"foo\": \"foobar\",".getBytes(StandardCharsets.US_ASCII);
+                    byte[] combined = Arrays.copyOf(ascii, ascii.length +  httpData.length() - 1);
+                    System.arraycopy(httpData.array(), 1, combined, ascii.length, httpData.length());
+                    return HttpData.wrap(combined);
+                }
+            } else {
+                // Http trailers.
+                assert obj instanceof HttpHeaders;
+            }
+            return obj;
```

Comment:
Just out of curiosity, would it be convenient if we provide
- `HttpResponse.mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> f)` 
- `HttpResponse.mapData(Function<? super HttpData, ? extends HttpData> f)`
- `HttpResponse.mapTrailers(Function<? super HttpHeaders, ? extends HttpHeaders> f)`

I think it would be convenient for stateless transforming. 
```java
HttpResponse response = ...;
response.mapHeaders(headers -> {
    if (headers.isInformational()) {
        return ob;
    } else {
        return headers.toBuilder()...build();
    }
}).mapData(data -> {
    ...
});
```

Reply:
Filed an issue for that. #3606

Reply:
I think the code in Armeria would benefit from this. I think most if not all of HttpDecodedResponse, SetCookieResponse and TruncatingHttpResponse could be removed in favour of those methods.

Reply:
This is super useful documentation -- chat pointed me to it and I used it recently to solve a problem I was fighting with.